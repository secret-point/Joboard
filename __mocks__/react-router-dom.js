// Mock navigate and useParams when needed
// const navigate = jest.fn();
// const useParams = jest.fn();
const useLocation = jest.fn();

module.exports = {
  ...jest.requireActual("react-router-dom"),
  // useNavigate: () => navigate,
  // useParams,
  useLocation,
};
