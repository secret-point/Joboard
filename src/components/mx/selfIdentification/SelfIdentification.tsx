import React, { useEffect, useState } from "react";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { useLocation } from "react-router-dom";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import queryString from "query-string";
import { ButtonVariant } from "@amzn/stencil-react-components/button";
import { Col, Row } from "@amzn/stencil-react-components/layout";
import { onCompleteTaskHelper } from "../../../actions/WorkflowActions/workflowActions";
import { CommonColors } from "../../../utils/colors";
import { Status, StatusIndicator } from "@amzn/stencil-react-components/status-indicator";
import { connect } from "react-redux";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";
import { JobState } from "../../../reducers/job.reducer";
import { boundResetBannerMessage } from "../../../actions/UiActions/boundUi";
import { boundUpdateApplicationDS } from "../../../actions/ApplicationActions/boundApplicationActions";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import { ApplicationState } from "../../../reducers/application.reducer";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { SelfIdentificationState } from "../../../reducers/selfIdentification.reducer";
import { UpdateApplicationRequestDS } from "../../../utils/apiTypes";
import {
  PROXY_APPLICATION_STATE,
  SELF_IDENTIFICATION_STEPS,
  UPDATE_APPLICATION_API_TYPE
} from "../../../utils/enums/common";
import {
  checkAndBoundGetApplication,
  createUpdateApplicationRequest, getCountryCode,
  getLocale,
  handleInitiateSelfIdentificationStep,
  isSelfIdentificationInfoValid,
  SelfShouldDisplayContinue
} from "../../../utils/helper";
import { translate as t } from "../../../utils/translator";
import { Application } from "../../../utils/types/common";
import InfoStepCard from "../../common/InfoStepCard";
import DisabilityForm from "../../common/self-Identification/disability-form";
import EqualOpportunityForm from "../../common/self-Identification/Equal-opportunity-form";
import VeteranStatusForm from "../../common/self-Identification/veteran-status-form";
import DebouncedButton from "../../common/DebouncedButton";

interface MapStateToProps {
  application: ApplicationState;
  candidate: CandidateState;
  selfIdentification: SelfIdentificationState;
  schedule: ScheduleState;
  job: JobState;
}

interface SelfIdentificationProps {

}

type SelfIdentificationMergeProps = MapStateToProps & SelfIdentificationProps;

export const SelfIdentification = (props: SelfIdentificationMergeProps) => {
  const { selfIdentification, application, candidate, schedule, job } = props;
  const { stepConfig } = selfIdentification;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId, scheduleId, jobId } = queryParams;
  const applicationData = application.results;
  const candidateData = candidate.results?.candidateData;
  const selfIdentificationInfo = candidateData?.selfIdentificationInfo;
  const [isSelfIdInfoValid, setIsSelfIdInfoValid] = useState(true);
  const jobDetail = job.results;
  const {scheduleDetail} = schedule.results;

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    boundGetCandidateInfo();
  }, [applicationData]);

  useEffect(() => {
    scheduleId && scheduleId!== scheduleDetail?.scheduleId && boundGetScheduleDetail({
      locale: getLocale(),
      scheduleId: scheduleId
    })
  }, [scheduleDetail, scheduleId]);

  useEffect(() => {
    jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() })
  }, [jobDetail, jobId]);

  useEffect(() => {
    jobDetail && applicationData && candidateData && scheduleDetail && addMetricForPageLoad(pageName);

  }, [jobDetail, applicationData, candidateData, scheduleDetail, pageName]);

  useEffect(() => {
    return () => {
      //reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    }
  },[pageName]);

  useEffect(() => {
    selfIdentificationInfo && handleInitiateSelfIdentificationStep(selfIdentificationInfo);
  }, [candidateData, applicationData, selfIdentificationInfo])


  const handleContinue = () => {
    boundResetBannerMessage();
    const isSelfIdInfoValid = isSelfIdentificationInfoValid(selfIdentificationInfo);
    setIsSelfIdInfoValid(isSelfIdInfoValid);

    if(applicationData && isSelfIdInfoValid) {
      const { SELF_IDENTIFICATION } = UPDATE_APPLICATION_API_TYPE;
      const payload = {
        state: PROXY_APPLICATION_STATE.SELF_IDENTIFICATION_COMPLETED
      }
      const request: UpdateApplicationRequestDS = createUpdateApplicationRequest(applicationData, SELF_IDENTIFICATION, payload);
      boundUpdateApplicationDS(request, (applicationData: Application)=>{
        onCompleteTaskHelper(applicationData);
      });
    }
  }

  const shouldRenderContinueButton = SelfShouldDisplayContinue(stepConfig);

  return (
    <Col gridGap={15}>
      <InfoStepCard
        title={t("BB-SelfId-equal-opportunity-form-title-text", "Voluntary Equal Opportunity Self-Identification Form")}
        expandedContent={<EqualOpportunityForm />}
        stepName={SELF_IDENTIFICATION_STEPS.EQUAL_OPPORTUNITY}
        infoCardStepStatus={stepConfig[SELF_IDENTIFICATION_STEPS.EQUAL_OPPORTUNITY]}
        stepIndex={1}
        maxStepNumber={2}
      />

      <InfoStepCard
        title={t("BB-SelfId-disability-form-card-title-text", "Voluntary Self-Identification of Disability")}
        expandedContent={<DisabilityForm />}
        stepName={SELF_IDENTIFICATION_STEPS.DISABILITY_FORM}
        infoCardStepStatus={stepConfig[SELF_IDENTIFICATION_STEPS.DISABILITY_FORM]}
        stepIndex={2}
        maxStepNumber={2}
      />
      {
        shouldRenderContinueButton && !isSelfIdInfoValid &&
        <Row padding="S300" backgroundColor={CommonColors.RED05}>
          <StatusIndicator
            messageText={"Please check all required boxes in previous steps to proceed."}
            status={Status.Negative}
            iconAriaHidden={true}
          />
        </Row>
      }
      <Col padding={{top: 'S300'}}>
        {
          shouldRenderContinueButton &&
          <DebouncedButton
            variant={ButtonVariant.Primary}
            onClick={handleContinue}
          >
            {t('BB-SelfId-form-continue-button-text', 'Continue')}
          </DebouncedButton>
        }
      </Col>
    </Col>
  );
};

const mapStateToProps = (state: MapStateToProps) => {
  return state;
};

export default connect(mapStateToProps)(SelfIdentification);
