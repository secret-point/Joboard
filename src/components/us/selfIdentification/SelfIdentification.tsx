import React, { useEffect, useState } from "react";
import { Col, Row } from "@amzn/stencil-react-components/layout";
import InfoStepCard from "../../common/InfoStepCard";
import { SELF_IDENTIFICATION_STEPS } from "../../../utils/enums/common";
import EqualOpportunityForm from "../../common/self-Identification/Equal-opportunity-form";
import { ApplicationState } from "../../../reducers/application.reducer";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { connect } from "react-redux";
import { SelfIdentificationState } from "../../../reducers/selfIdentification.reducer";
import VeteranStatusForm from "../../common/self-Identification/veteran-status-form";
import DisabilityForm from "../../common/self-Identification/disability-form";
import {
  checkAndBoundGetApplication, getLocale,
  handleInitiateSelfIdentificationStep,
  isSelfIdentificationInfoValid,
  SelfShouldDisplayContinue
} from "../../../utils/helper";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { useLocation } from "react-router";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import queryString from "query-string";
import { Button, ButtonVariant } from "@amzn/stencil-react-components/button";
import { translate as t } from "../../../utils/translator";
import { onCompleteTaskHelper } from "../../../actions/WorkflowActions/workflowActions";
import { CommonColors } from "../../../utils/colors";
import { Status, StatusIndicator } from "@amzn/stencil-react-components/status-indicator";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";
import { JobState } from "../../../reducers/job.reducer";

interface MapStateToProps {
  application: ApplicationState,
  candidate: CandidateState,
  selfIdentification: SelfIdentificationState,
  schedule: ScheduleState,
  job: JobState;
}

interface SelfIdentificationProps {

}

type SelfIdentificationMergeProps = MapStateToProps & SelfIdentificationProps;

const SelfIdentificationComponent = (props: SelfIdentificationMergeProps) => {
  const { selfIdentification, application, candidate, schedule, job } = props;
  const { stepConfig } = selfIdentification;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId, scheduleId, jobId } = queryParams;
  const applicationData = application.results;
  const candidateData = candidate.results?.candidateData;
  const selfIdentificationInfo = candidateData?.selfIdentificationInfo;
  const [isSelfIdInfoValid, setIsSelfIdInfoValid] = useState(true);
  const jobDetail = job.results;
  const scheduleDetail = schedule.results.scheduleDetail;

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    boundGetCandidateInfo();
  }, [applicationData]);

  useEffect(() => {
    scheduleId && scheduleId!== scheduleDetail?.scheduleId && boundGetScheduleDetail({
      locale: getLocale(),
      scheduleId: scheduleId
    })
  }, [scheduleId]);

  useEffect(() => {
    jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() })
  }, [jobDetail, jobId]);

  useEffect(() => {
    jobDetail && applicationData && candidateData && scheduleDetail && addMetricForPageLoad(pageName);

  }, [jobDetail, applicationData, candidateData, scheduleDetail]);

  useEffect(() => {
    return () => {
      //reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    }
  },[])

  useEffect(() => {
    selfIdentificationInfo && handleInitiateSelfIdentificationStep(selfIdentificationInfo);
  }, [candidateData, applicationData, selfIdentificationInfo])


  const handleContinue = () => {
    const isSelfIdInfoValid = isSelfIdentificationInfoValid(selfIdentificationInfo);
    setIsSelfIdInfoValid(isSelfIdInfoValid);

    if(applicationData && isSelfIdInfoValid) {
      onCompleteTaskHelper(applicationData);
    }
  }

  const shouldRenderContinueButton = SelfShouldDisplayContinue(stepConfig);

  return (
    <Col gridGap={15}>
      <InfoStepCard
        title={t("BB-SelfId-equal-opportunity-form-title-text", "Voluntary Equal Opportunity Self-Identification Form")}
        expandedContent={<EqualOpportunityForm />}
        stepName={SELF_IDENTIFICATION_STEPS.EQUAL_OPPORTUNITY}
        infoCardStepStatus={stepConfig[SELF_IDENTIFICATION_STEPS.EQUAL_OPPORTUNITY]}
        stepIndex={1}
      />

      <InfoStepCard
        title={t("BB-SelfId-veteran-status-title-text", "Voluntary Self-Identification of Veteran Status")}
        expandedContent={<VeteranStatusForm />}
        stepName={SELF_IDENTIFICATION_STEPS.VETERAN_FORM}
        infoCardStepStatus={stepConfig[SELF_IDENTIFICATION_STEPS.VETERAN_FORM]}
        stepIndex={2}
      />

      <InfoStepCard
        title={t("BB-SelfId-disability-form-card-title-text", "Voluntary Self-Identification of Disability")}
        expandedContent={<DisabilityForm />}
        stepName={SELF_IDENTIFICATION_STEPS.DISABILITY_FORM}
        infoCardStepStatus={stepConfig[SELF_IDENTIFICATION_STEPS.DISABILITY_FORM]}
        stepIndex={3}
      />
      {
        shouldRenderContinueButton && !isSelfIdInfoValid &&
        <Row padding="S300" backgroundColor={CommonColors.RED05}>
          <StatusIndicator
            messageText={"Please check all required boxes in previous steps to proceed."}
            status={Status.Negative}
            iconAriaHidden={true}
          />
        </Row>
      }
      <Col padding={{top: 'S300'}}>
        {
          shouldRenderContinueButton &&
          <Button
            variant={ButtonVariant.Primary}
            onClick={handleContinue}
          >
            {t('BB-SelfId-form-continue-button-text', 'Continue')}
          </Button>
        }
      </Col>
    </Col>
  );
};

const mapStateToProps = (state: MapStateToProps) => {
  return state;
};

export default connect(mapStateToProps)(SelfIdentificationComponent);
