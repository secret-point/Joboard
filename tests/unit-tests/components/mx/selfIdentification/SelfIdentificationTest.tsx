import React from "react";
import { shallow } from "enzyme";
import { useLocation } from "react-router-dom";
import { SelfIdentification } from "../../../../../src/components/mx/selfIdentification/SelfIdentification";
import {
  TEST_APPLICATION_ID,
  TEST_APPLICATION_STATE,
  TEST_CANDIDATE_STATE,
  TEST_JOB_ID,
  TEST_JOB_STATE,
  TEST_SCHEDULE_ID,
  TEST_SCHEDULE_STATE,
  TEST_SELF_IDENTIFICATION_STATE
} from "../../../../test-utils/test-data";

describe("SelfIdentification", () => {
  const mockLocation = {
    pathname: "/self-identification",
    search: `?jobId=${TEST_JOB_ID}&applicationId=${TEST_APPLICATION_ID}&scheduleId=${TEST_SCHEDULE_ID}`,
    hash: '',
    state: null
  };
  const mockUseLocation = useLocation as jest.Mock;
  mockUseLocation.mockReturnValue(mockLocation);

  it("should match snapshot", () => {

    const shallowWrapper = shallow(
      <SelfIdentification
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
        selfIdentification={TEST_SELF_IDENTIFICATION_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });
});
