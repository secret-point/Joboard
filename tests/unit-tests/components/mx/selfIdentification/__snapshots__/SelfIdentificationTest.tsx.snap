// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelfIdentification should match snapshot 1`] = `
<Col
  gridGap={15}
>
  <Connect(InfoStepCard)
    expandedContent={<Memo(Connect(EqualOpportunityForm)) />}
    infoCardStepStatus={
      Object {
        "editMode": false,
        "status": "ACTIVE",
      }
    }
    maxStepNumber={2}
    stepIndex={1}
    stepName="EQUAL_OPPORTUNITY_FORM"
    title="Voluntary Equal Opportunity Self-Identification Form"
  />
  <Connect(InfoStepCard)
    expandedContent={<Memo(Connect(DisabilityForm)) />}
    infoCardStepStatus={
      Object {
        "editMode": true,
        "status": "ACTIVE",
      }
    }
    maxStepNumber={2}
    stepIndex={2}
    stepName="DISABILITY_FORM"
    title="Voluntary Self-Identification of Disability"
  />
  <Col
    padding={
      Object {
        "top": "S300",
      }
    }
  />
</Col>
`;
